### PR TITLE
calculating stake distribution with relations

### PR DESCRIPTION
**Pull request by JaredCorduan**
This PR presents the stake distribution calculation in a different way, by using relations:

relations

staked

Additionally, I've removed the `PooledStake` type, and no longer pass around that nested structure (which consolidated the pool owners into a single key value pair).  Now instead we pass along the (un-pooled) stake distribution together with the delegation map. Such as in the stake distribution snapshots:

pstake

When we do need to filter down to one pool, we use this new method:

poolstake

Besides that, there were just a few minor edits:
1) there was still a reference to the `ACNT` rule, which is now called `REWARD`
2) there was a paragraph about how withdrawal accounts work that was describing how things _used_ to work, so it was removed
3) a clarification in the prose where "delegation certificate" was vague (ie all five types or just the one type that delegates stake to a pool).

# UPDATE

I've redone the relation notation.  

We now freely think of maps and functions as relations, and define inverses and composition:

relations

Then the two tables for the relations and the actual stake distribution are now just one table:

newstakedistr

Note that the `aggregate_+` function is now generic.

Finally, I've tried to break down the final Addr to Coin relation in the prose:

foo

### CODE REVIEW

**Comment by dnadales** [comment]
Any reason why we have assignee's instead of reviewers this time? Trying to make it hard for people to ignore your PR's :P :)

**Comment by JaredCorduan** [comment]
> Any reason why we have assignee's instead of reviewers this time? Trying to make it hard for people to ignore your PR's :P :)

Oops, I did not mean to do that! :) Thank you for letting me know!

File: latex/epoch.tex [hunk 1]
```diff
@@ -232,74 +231,59 @@ \subsection{Stake Distribution Calculation}
...
+        (ptrs,~hk,~a) & \fun{ptrRel} & \fun{addrPtr}~a\mapsto hk\in\var{ptrs} \\
+        (\wcard,~hk,~a) & \fun{baseRel} & \fun{stakeHK_b}~a=hk \\
+        (\wcard,~hk,~a) & \fun{rewRel} & \fun{stakeHK_r}~a=hk \\
+      \end{array}
+    \end{equation*}
+    \begin{equation*}
+      \fun{hkRel} = \fun{baseRel}\cup\fun{ptrRel}\cup\fun{rewRel}
+    \end{equation*}
+
+  \emph{Aggreation}
```
Comment (dnadales) [inline latex/epoch.tex]:
```suggestion
  \emph{Aggregation}
```

File: latex/epoch.tex [hunk 2]
```diff
@@ -315,25 +299,13 @@ \subsection{Stake Distribution Calculation}
       & \where \\
       & ~~~~ (\var{stkeys},~\var{rewards},~\var{delegations},~\var{ptrs}) = \var{dstate} \\
       & ~~~~ (\var{stpools},~\wcard,~\wcard,~\wcard) = \var{pstate} \\
-      & ~~~~ \var{outs} = \fun{consolidate}~{utxo} \\
-      & ~~~~ \var{stake} = (\fun{baseStake}~{outs})
-                             \unionoverridePlus (\fun{ptrStake}~{outs}~{ptrs})
-                             \unionoverridePlus (\fun{rewardStake}~{rewards}) \\
+      & ~~~~ \var{stake} = \fun{aggregate_{+}}~
+        \left(\fun{utxoRel}~\var{utxo}\right)\circ\left(\fun{hkRel}~\var{ptrs}\right) \\
```
Comment (dnadales) [inline latex/epoch.tex]: Did we introduce function composition? I guess we take the definition in which the first function being applied is to the right of the composition symbol ...

File: latex/epoch.tex [hunk 3]
```diff
@@ -315,25 +299,13 @@ \subsection{Stake Distribution Calculation}
       & \where \\
       & ~~~~ (\var{stkeys},~\var{rewards},~\var{delegations},~\var{ptrs}) = \var{dstate} \\
       & ~~~~ (\var{stpools},~\wcard,~\wcard,~\wcard) = \var{pstate} \\
-      & ~~~~ \var{outs} = \fun{consolidate}~{utxo} \\
-      & ~~~~ \var{stake} = (\fun{baseStake}~{outs})
-                             \unionoverridePlus (\fun{ptrStake}~{outs}~{ptrs})
-                             \unionoverridePlus (\fun{rewardStake}~{rewards}) \\
+      & ~~~~ \var{stake} = \fun{aggregate_{+}}~
```
Comment (dnadales) [inline latex/epoch.tex]: I think some parenthesis might help in reading this.

File: latex/epoch.tex (hunk 2 above)
Comment (JaredCorduan) [reply latex/epoch.tex]: ah yes, thank you, I forgot to define composition of relations.

**Comment by JaredCorduan** [comment]
@dnadales thank you again for your help today! I've redone the notation for relations. I've added an `UPDATE` section to the PR description above.

**Pull request merged**